rewriting BarCodeScanning Functionality

fabriziobertoglio1987 opened the pull request:
<!-- Thanks for submitting a pull request! We appreciate you spending the time to work on these changes. Please follow the template so that the reviewers can easily understand what the code changes affect -->

# Summary

@bqwang91 Rewriting the logic responsible for limiting the scanning area for Barcodes 

1) Deleted https://github.com/react-native-community/react-native-camera/pull/2719 and previous logic
2) Built [BarCodeScanner][5] API based on the [`zxing`][3] API. The `rectOfInterest` prop is replaced from `planarLuminanceSourceParams`. IOs can apply the same changes in a separate pull request as it is using the same library (zxing) or keep the `rectOfInterest` prop.
3) Adding a separate Javascript Implementation Example (separate pull request)

The motivations for this design are [described here][8].
A simple [**JS implementation**][9].

<img src="https://user-images.githubusercontent.com/24992535/78683445-975a0f00-78ef-11ea-988a-0bdfe82433de.gif" alt="alt text" width="200px" height="whatever">

[5]: https://github.com/fabriziobertoglio1987/react-native-camera/blob/encapsulating-logic-in-area-class/example/examples/mlkit/README.markdown#table-of-contents
[3]: https://zxing.github.io/zxing/apidocs/com/google/zxing/PlanarYUVLuminanceSource.html
[8]: https://github.com/react-native-community/react-native-camera/pull/2766#issuecomment-605499719
[9]: https://github.com/fabriziobertoglio1987/react-native-camera/tree/encapsulating-logic-in-area-class/example/examples/mlkit#table-of-contents

## Test Plan

<!-- Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots / videos if the pull request changes UI. -->

### What's required for testing (prerequisites)?
The functionality was tested with the front camera in Portrait and Landscape Mode on devices with different screen sizes.
I still have to test the functionality compatability with the GoogleBarcode functionality, but I see no reason for it to break as it is using the [BarcodeDetectorAsyncTask][7].

[7]: https://github.com/react-native-community/react-native-camera/blob/c4eb697045c424a414ae82b68920dbe020d7b9ce/android/src/main/java/org/reactnative/camera/RNCameraView.java#L188-L189

### What are the steps to reproduce (after prerequisites)?
The mlkit example to test this functionality is included in [this branch][9] of my fork, but the Example is yet to be completed.

[2]: https://github.com/fabriziobertoglio1987/react-native-camera/tree/encapsulating-logic-in-area-class/example

You open the camera and scan one of your barcode, the value is outputted in your server log.
You need to run `yarn react-native log-android` to display the barcode code outputted in the server log.

## Compatibility

| OS      | Implemented |
| ------- | :---------: |
| iOS     |    ❌     |
| Android |    ✅     |

## Checklist

<!-- Check completed item, when applicable, via: [X] -->

- [X] I have tested this on a device and a simulator
- [ ] I added the documentation in `README.md`
- [ ] I mentioned this change in `CHANGELOG.md`
- [ ] I updated the typed files (TS and Flow)
- [X] I added a sample use of the API in the example project (`example/App.js`)

SimonErm commented:
Looks good to me. But is there any reason, why you use HashMaps for height/width and x/y instead of https://developer.android.com/reference/android/util/Size and Point?

fabriziobertoglio1987 commented on android/src/main/java/org/reactnative/camera/tasks/BarCodeScannerAsyncTask.java:
```diff
@@ -60,81 +38,32 @@ protected Result doInBackground(Void... ignored) {
+      scanWidth,
+      false
+    );
+    BinaryBitmap bitmap = new BinaryBitmap(new HybridBinarizer(source));
```
@sibelius @bqwang Thanks a lot. I believe the logic is still very complex, in fact I'm reading the [PlanarYUVLuminanceSource][1] documentation and I notice now that the class is pretty simple

```java
PlanarYUVLuminanceSource(
     byte[] yuvData, // the image taken
     int dataWidth, // the width of the original picture containing barcode
     int dataHeight, // the heigh of the original picture containing barcode
     // BELOW are the cropping parameters, requires x,y starting point and width/height
     int left, // x coordinates of the top left corner starting point
     int top,  // y coordinates of the top left corner starting point
     int width, // width to crop image in this direction --->>>>
     int height, // height to crop image in this direction    V
     boolean reverseHorizontal
) 
```
we specify the left top corner, width and height for cropping. Seems to me pretty simple api and `rectOfInterest` should just pass those 4 parameters as specified below. (`x` is `left`, `y` is `top`)

```javascript
rectOfInterest={{ x: scanAreaX, y: scanAreaY, width: scanAreaWidth, height: scanAreaHeight }}
```

I believe `react-native-camera` should just expose the [zxing][2] to the `Javascript` user. The user should handle the rest. The advanteges would be:

1) We base our API on the official Android/Java apis/docs
2) We have less code
3) We have less bugs and manteinance burden. Is the responsability of the developer to handle all the remaning logic.

Took me some time to figure this out, but your comment helped me as I felt that I could not make any more improvements excepts from a totally new starting point. 

This means 90% of the previous functionality needs to be deleted, because the entire logic inside the [ScanArea](https://github.com/react-native-community/react-native-camera/blob/b258e447a8a9baf6a7c899d289c2534c53c88745/android/src/main/java/org/reactnative/camera/utils/ScanArea.java) class is just built on top of [PlanarYUVLuminanceSource][1] and superflous. 

Thanks a lot

[1]: https://zxing.github.io/zxing/apidocs/com/google/zxing/PlanarYUVLuminanceSource.html
[2]: https://github.com/zxing/zxing

fabriziobertoglio1987 replied:
This works with my phone. `mImageData` is the original portrait image rotated to landscape (see `rotateImage` method bottom of the page).
```java
mImageData = rotateImage(imageData,mWidth, mHeight);
```
this explains why dataWidth and dataHeight are inverted (as if the image was landscape instead of portrait). The same applies for left, top, width and height cropping parameters.

```java
    PlanarYUVLuminanceSource source = new PlanarYUVLuminanceSource(
      mImageData, // byte[] yuvData, the image taken
      1440, // int dataWidth, the width of the original picture containing barcode
      1080, // int dataHeight, the heigh of the original picture containing barcode
       // BELOW are the cropping parameters, requires x,y starting point and width/height
       266, // int left, x coordinates of the top left corner starting point
       192, // int top,  y coordinates of the top left corner starting point
       533, // int width, width to crop image in this direction --->>>>
       666, // int height, height to crop image in this direction    V
       false, // boolean reverseHorizontal
    );
```
The `rotateImage()` method flips the image from portrait to landscape.

```java 

private byte[] rotateImage(byte[]imageData,int width, int height) {
  byte[] rotated = new byte[imageData.length];
  for (int y = 0; y < height; y++) {
    for (int x = 0; x < width; x++) {
      rotated[x * height + height - y - 1] = imageData[x + y * width];
    }
  }
  return rotated;
}
```

fabriziobertoglio1987 replied:
I took [a picture][1] with the Camera App and opened with Image Editing Tools (gimp).
The `this.camera.takePictureAsync()` callback returns `data` with height 4160 and width 2080 pixels. 

I [converted](https://www.freefileconvert.com/jpg-yuv) the picture to `YUV` format

I display the image on https://rawpixels.net/

1) browse and choose the yuv file
2) set width 2080 and height 4160

The image shows exactly like the original but distorted colors.

Now I am inspecting the pixel using [`getMatrix`](https://zxing.github.io/zxing/apidocs/com/google/zxing/PlanarYUVLuminanceSource.html#getMatrix--)

[1]: https://github.com/react-native-community/react-native-camera/pull/2766

fabriziobertoglio1987 replied:
zxing library includes [a demo](https://github.com/zxing/zxing/blob/a65631d0b643e778b8b1c1975e4f18f35e401fa4/android/src/com/google/zxing/client/android/camera/CameraManager.java#L319-L325) for this functionality

fabriziobertoglio1987 replied:
Test performed with another phone in lascape mode. The image data is **not rotated**.
```java
     PlanarYUVLuminanceSource source = new PlanarYUVLuminanceSource(
       mImageData,
       1520,
       1140,
       400,
       400,
       800,
       400, 
       false
    );
```

fabriziobertoglio1987 replied:
Portrait 
```java
     PlanarYUVLuminanceSource source = new PlanarYUVLuminanceSource(
       rotateImage(mImageData, mWidth, mHeight),
       height, 
       width,
       top,
       left,
       height,
       width,
       false
    );
```
Landscape
```java
     PlanarYUVLuminanceSource source = new PlanarYUVLuminanceSource(
       mImageData,
       width,
       height, 
       left,
       top,
       width,
       height,
       false
    );
```

fabriziobertoglio1987 commented:
Thanks @sibelius I would need your advise and preference between **OPTION A** and **OPTION B**

**SHORT VERSION OF THIS POST**
**OPTION A** - Improve the **CURRENT SOLUTION**  without removing the [`ScanArea`][5] logic
Adding Functionalities on top of [`zxing`][4] library to make Barcode Scanner easier to implement for a React Developer.

**ADVANTAGES**
- It may be faster/easier to implement Barcode Scanner for `ReactNative` developer

**DISADVANTAGES**
- We could use an android package based on `zxing` instead of implementing our own  solution based on `zxing`. No need to reinvent the wheel.
- The logic we wrote on top of `zxing` is hard to mantain
- Lack of documentation on how to use our functionality
- Different apps may need different solutions and our implementation may not work well for all use cases. For example our use case may work in Portrait, but may not work well in Lanscape mode or with the [Back Camera](https://stackoverflow.com/questions/55846066/front-facing-camera-does-not-auto-focus-or-manually-focus). Given the large usage of the libraries, there are a lot of use cases to take in consideration.

[4]: https://github.com/zxing/zxing
[5]: https://github.com/react-native-community/react-native-camera/blob/b258e447a8a9baf6a7c899d289c2534c53c88745/android/src/main/java/org/reactnative/camera/utils/ScanArea.java

**OPTION B** - Remove the [`ScanArea`][5] old logic and expose the `zxing` api to the `ReactNative` developer, including a separate example in the wiki on how to implement the functionality and reference to the original `zxing` project.

**ADVANTAGES**
- The **DISADVANTAGES** of **Option A**
- The `zxing` [demo app][1] seems to work **very** well
- We could later add more android libraries and expose them to the `ReactNative` developer via the bridge. The other libraries will make it easier to implement Barcode scanning
- Our functionality will be very simple and it will never break. It will consist in this 10 lines of code.
```java
    PlanarYUVLuminanceSource source = new PlanarYUVLuminanceSource(
      mImageData, // byte[] yuvData, the image taken
      1440, // int dataWidth, the width of the original picture containing barcode
      1080, // int dataHeight, the heigh of the original picture containing barcode
       266, // int left, x coordinates of the top left corner starting point
       192, // int top,  y coordinates of the top left corner starting point
       533, // int width, width to crop image in this direction --->>>>
       666, // int height, height to crop image in this direction    V
       false, // boolean reverseHorizontal
    );
```
- It will be easy to test the above code, for ex:
UNIT TEST 1) expect any instance of `PlanarYUVLuminanceSource` class to receive the `new` method with parameters `mImageData`, `dataWidth`, `dataHeight` .. and return ...
UNIT TEST 2) expect that any other method passes the correct arguments and raises the correct errors
- We will include an implementation example and we could build a separate package in the future

**DISADVANTAGES**
- Javascript logic implemented from client side User may be slower

**LONG VERSION OF THIS POST WITH TECHNICAL DETAILS**

Based on the [`zxing android`](https://github.com/zxing/zxing/tree/a65631d0b643e778b8b1c1975e4f18f35e401fa4/android) Demo SourceCode I can implement this solution in ReactNative:

1) The method [`setManualFramingRect()`][2] called from [`buildLuminanceSource()`][3] calculates a Reactangle.
2) The `reactangle` is calculated based on the [screen resolution](https://developer.android.com/reference/android/view/WindowManager#getCurrentWindowMetrics()) and original image `width` and `height`. 
3) [`buildLuminanceSource()`][3] calls [`PlanarYUVLuminanceSource`](https://zxing.github.io/zxing/apidocs/com/google/zxing/PlanarYUVLuminanceSource.html) to crop and scan the image.

```java
private static final int MIN_FRAME_WIDTH = 240;
private static final int MIN_FRAME_HEIGHT = 240;
private static final int MAX_FRAME_WIDTH = 1200; // = 5/8 * 1920
private static final int MAX_FRAME_HEIGHT = 675; // = 5/8 * 1080

public synchronized Rect getFramingRect() {
  if (framingRect == null) {
    if (camera == null) {
      return null;
    }
    Point screenResolution = configManager.getScreenResolution();
    if (screenResolution == null) {
      // Called early, before init even finished
      return null;
    }

    int width = findDesiredDimensionInRange(screenResolution.x, MIN_FRAME_WIDTH, MAX_FRAME_WIDTH);
    int height = findDesiredDimensionInRange(screenResolution.y, MIN_FRAME_HEIGHT, MAX_FRAME_HEIGHT);

    int leftOffset = (screenResolution.x - width) / 2;
    int topOffset = (screenResolution.y - height) / 2;
    framingRect = new Rect(leftOffset, topOffset, leftOffset + width, topOffset + height);
    Log.d(TAG, "Calculated framing rect: " + framingRect);
  }
  return framingRect;
}

private static int findDesiredDimensionInRange(int resolution, int hardMin, int hardMax) {
  int dim = 5 * resolution / 8; // Target 5/8 of each dimension
  if (dim < hardMin) {
    return hardMin;
  }
  return Math.min(dim, hardMax);
}

/** * A factory method to build the appropriate LuminanceSource object based on the format
* of the preview buffers, as described by Camera.Parameters.
 *
 * @param data A preview frame.
 * @param width The width of the image.
 * @param height The height of the image.
 * @return A PlanarYUVLuminanceSource instance.
 */
public PlanarYUVLuminanceSource buildLuminanceSource(byte[] data, int width, int height) {
  Rect rect = getFramingRectInPreview();
  if (rect == null) {
    return null;
  }
  // Go ahead and assume it's YUV rather than die.
  return new PlanarYUVLuminanceSource(data, width, height, rect.left, rect.top,
                                      rect.width(), rect.height(), false);
}

```
Thanks a lot Sibelius. I wish you a good weekend and a good evening.
Fabrizio

[1]: https://play.google.com/store/apps/details?id=com.google.zxing.client.android
[2]: https://github.com/zxing/zxing/blob/a65631d0b643e778b8b1c1975e4f18f35e401fa4/android/src/com/google/zxing/client/android/camera/CameraManager.java#L289-L307
[3]: https://github.com/zxing/zxing/blob/a65631d0b643e778b8b1c1975e4f18f35e401fa4/android/src/com/google/zxing/client/android/camera/CameraManager.java#L318-L326

sibelius commented:
Can you show the JavaScript code of option B? Can we make a hook or expose some helper function to make this easy on user land?

fabriziobertoglio1987 commented:
@bqwang91 

I found out that the Camera Resolution is [the quality used](https://github.com/react-native-community/react-native-camera/blob/master/docs/RNCamera.md#defaultvideoquality) and it does not relate to the `screen` resolution. I can preview/display the video at *1080p* (1920x1080 pixel), but record at *720p* (1280 x 720pixel). The previous approach below need to be replaced with a method which uses the react native bridge to retrieve the current resolution from the camera.

```javascript
const CAM_VIEW_HEIGHT = Dimensions.get('screen').width * 1.5;
const CAM_VIEW_WIDTH = Dimensions.get('screen').width;
```

The aspect ratio may also be involved to avoid distorsions, in the `zxing` android app the method [findBestPreviewSizeValue][1] select the correct resolution.

I am currently searching how to retrieve the camera resolution via the `react-native-api` or I may have to create a new method...

[1]: https://github.com/zxing/zxing/blob/0cf3b9be71680f50c90a71ca26ce0d33664b0dd6/android-core/src/main/java/com/google/zxing/client/android/camera/CameraConfigurationUtils.java#L273-L345

fabriziobertoglio1987 commented on android/src/main/java/org/reactnative/camera/CameraViewManager.java:
```diff
@@ -201,25 +201,16 @@ public void setTextRecognizing(RNCameraView view, boolean textRecognizerEnabled)
+      view.setRectOfInterest(dataWidth, dataHeight, left, top, width, height);
     }
   }
-  /**---limit scan area addition---**/
```
`dataWidth` and `dataHeight` should be exposed to the javascript developer with a `getter`. 
`cameraView.getPreviewSize` returns the `dataWidth` and `dataHeight`, while `cameraView.getPictureSize` returns the image height and width.

fabriziobertoglio1987 commented on android/src/main/java/org/reactnative/camera/CameraViewManager.java:
```diff
@@ -201,25 +201,16 @@ public void setTextRecognizing(RNCameraView view, boolean textRecognizerEnabled)
-      view.setCameraViewDimensions(cameraViewWidth, cameraViewHeight);
+      int dataWidth = (int) coordinates.getInt("dataWidth");
+      int dataHeight = (int) coordinates.getInt("dataHeight");
+      int left = (int) coordinates.getInt("left");
```
`dataWidth` and `dataHeight` should be exposed to the javascript developer with a `getter`. 
`cameraView.getPreviewSize` returns the `dataWidth` and `dataHeight`, while `cameraView.getPictureSize` returns the image height and width.

fabriziobertoglio1987 commented:
@sibelius It is still work in progress, I updated the [summary](https://github.com/react-native-community/react-native-camera/pull/2766#issue-393843055) of this pull request and added an simple javascript implementation using hooks on a separate branch.

The example is work in progress, does not take in consideration device orientation (by default landscape) or other aspects (like image distorsions caused by resizing images from the screen preview aspect ratio to the camera recording aspect ratio), but all information are now available to the `React Native` developer which can build this functionality for his use case and contribute to our example of implementation.

I included some explanation in the [`README`](https://github.com/fabriziobertoglio1987/react-native-camera/tree/encapsulating-logic-in-area-class/example/examples/mlkit#rectangle-of-interest-for-barcode-scanning-android) on a separate branch. It is Work In Progress and I will keep improving it tomorrow. 

Thanks a lot. Fabrizio

[1]: https://github.com/fabriziobertoglio1987/react-native-camera/tree/encapsulating-logic-in-area-class/example/examples/mlkit#rectangle-of-interest-for-barcode-scanning-android

fabriziobertoglio1987 commented on android/src/main/java/org/reactnative/camera/tasks/BarCodeScannerAsyncTask.java:
```diff
@@ -2,139 +2,66 @@
-    int scanWidth = (int) (mScanAreaWidth * mWidth);
-    int scanHeight = (int) (((mScanAreaHeight * mCameraViewWidth) / adjustedCamViewWidth) * mHeight);
-
+    boolean landscapeMode = mDataWidth > mDataHeight;
```
I forgot to unlock my phone orientation settings and I will have to make some modification.

fabriziobertoglio1987 commented:
I am very happy to say that this feature is perfectly working and completed. I wrote a clear explanation on how to implement the [javascript functionality](https://github.com/fabriziobertoglio1987/react-native-camera/tree/encapsulating-logic-in-area-class/example/examples/mlkit#visual-explanation), including working example of the functionality and clear explanation to understand on how does the original api work....

I will prepare next week a pull request to integrate and add this documentation to our repository...

I am happy to bring further improvements ... Thanks a lot everyone for the support ... Let's hope for the best and I wish you a good weekend to everyone.

sibelius commented:
How do I set rectOfInterest in this approach?

Is this as fast as the other approach?

fabriziobertoglio1987 commented:
Mabye I can try to build a custom hook for this [**javascript implementation**](https://github.com/fabriziobertoglio1987/react-native-camera/tree/encapsulating-logic-in-area-class/example/examples/mlkit#javascript-implementation), but I need `onCameraReady` and `getCameraSettings()` hooks ...

so I'll probably need to set this pull request aside and work on those tasks...

fabriziobertoglio1987 commented on android/src/main/java/org/reactnative/camera/RNCameraView.java:
```diff
@@ -356,17 +355,10 @@ public void onBarCodeScanningTaskCompleted() {
-    public void setCameraViewDimensions(int width, int height) {
-    this.mCameraViewWidth = width;
-    this.mCameraViewHeight = height;
+  public void setRectOfInterest(int dataWidth, int dataHeight, int left, int top, int width, int height) {
```
I should test for conflicts with IOS functioanality and consider using a different name for this prop...

fabriziobertoglio1987 replied:
I need to add a type [here](https://github.com/react-native-community/react-native-camera/blob/5abbacaac4448311775c5544ea5feb3c3b5b94f9/src/RNCamera.js#L285) or change the existing `rectOfInterest`. It is still not clear to me if it is already merged for iOS.

sibelius replied:
so we gonna have both props?

fabriziobertoglio1987 replied:
@sibelius I am investigating if the functionality was implemented for iOS. Additionally I had made some small changes to the implementation example https://github.com/fabriziobertoglio1987/react-native-camera/commit/ea167f6b8f33ecc37772e618c273d0ae74da68dc to avoid breaking the ios functionalities.

I see many non-merged pull request for this functionality, so maybe it was not implemented for ios. Searching [rectOfInterest](https://github.com/react-native-community/react-native-camera/search?q=rectOfInterest&unscoped_q=rectOfInterest) and [setRectOfInterest](https://github.com/react-native-community/react-native-camera/search?q=setRectOfInterest&unscoped_q=setRectOfInterest) only shows functionality to store the value, but not for using it in `zxing`...

I will check out again https://github.com/react-native-community/react-native-camera/pull/2719 and test that that it works also on iOS and keep searcing...

Maybe rectOfInterest will be removed... Thanks a lot

fabriziobertoglio1987 replied:
https://github.com/react-native-community/react-native-camera/pull/2209

fabriziobertoglio1987 replied:
Hopefully in the android code I can detect the screen orientation and pass the props as `float` value from 0 to 1 (a percentage relative to the screen dimension) as in the ios functioanlity and have `rectOnProps` be exactly the same between iOS and Android

sibelius commented:
@cristianoccazinsp can you take a look of this approach?

cristianoccazinsp commented:
The changes look good to me, although the discussion/motivation is a bit difficult to understand.

What's the final goal of this change?

Me, as a developer updating RNCamera, what should I change if I'm already using the current barcode scanner behaviour? What should I do if I want to use this improvement?

Might want to have a brief and concise description to be included in the release notes.

fabriziobertoglio1987 commented:
Thanks @cristianoccazinsp

> the discussion/motivation is a bit difficult to understand.
> What's the final goal of this change?

This pull request is a follow up of [the following discussion](https://github.com/react-native-community/react-native-camera/pull/2719#discussion_r391133195).

> as a developer updating RNCamera, what should I change if I'm already using the current barcode scanner behaviour? What should I do if I want to use this improvement?

I am still working on this functionality, maybe it will work exactly as on IOS and no change may be required, **if the user implemented correctly the functionality**, but we did not either verify that https://github.com/react-native-community/react-native-camera/pull/2719 actually worked, so we can not tell if this pull request is building the entire functionality from scratch or if the previous pull request did not work.

We did not assert.

fabriziobertoglio1987 commented:
Thanks @cristianoccazinsp

> the discussion/motivation is a bit difficult to understand.
> What's the final goal of this change?

This pull request is a follow up of this [#2719 comment](https://github.com/react-native-community/react-native-camera/pull/2719#discussion_r391133195) . It is a re-write of the original functionality.

> as a developer updating RNCamera, what should I change if I'm already using the current barcode scanner behaviour? What should I do if I want to use this improvement?

As a Company working on RNCamera that needs the `rectOfInterest` functionality on their `Android` application, they would have to read the [**instructions**](https://github.com/fabriziobertoglio1987/react-native-camera/tree/encapsulating-logic-in-area-class/example/examples/mlkit#table-of-contents) to implement the functionality on Android and use a different prop called [`planarLuminanceSourceParams`](https://github.com/react-native-community/react-native-camera/pull/2766/commits/5abbacaac4448311775c5544ea5feb3c3b5b94f9#diff-5236020b94a21b26a9b70c84bbe86dbdR205) to set the `rectOfInterest` for android devices.

The Company would invest some time and implement a new BarCode functionality in their Mobile Android Application, otherwise I opened feature https://github.com/react-native-community/react-native-camera/issues/2783 for further integration of the `rectOfInterest` prop between iOS and Android (see https://github.com/react-native-community/react-native-camera/pull/2766#discussion_r405545196). 

@sibelius from today I will not work anymore on this pull request or the `react-native-camera` project. I will invest time and contribute to [tfjs/tfjs-react-native](https://github.com/tensorflow/tfjs/tree/master/tfjs-react-native). 

Yes, Android and iOS will use two different props (for now), as I did not have enough time to develop https://github.com/react-native-community/react-native-camera/issues/2783

@sibelius @cristianocca https://github.com/react-native-community/react-native-camera/pull/2719 was merged 27 days ago, only one Company is using the functionality and they probably can go back using their own fork while we keep improving our functionality.

Let me know if you want me to work on https://github.com/react-native-community/react-native-camera/issues/2783

I need your final approvale on merging this code, then I will then run final test, fix any bug and add brief readme/changelog. 

I worked on this for 20 days. It is a full rewrite. I have very good understanding of this functionality and I am 100% capable of delivering the best improvements with the best solution.

https://fabriziobertoglio.xyz

Thanks a lot.
Fabrizio
